helix-view/src/register.rs

```diff
@@ -0,0 +1,235 @@
+            '#' | '.' | '%' => Err(anyhow::anyhow!("Register {name} does not support pushing")),
+            '*' | '+' => {
+                self.inner.entry(name).or_insert_with(Vec::new).push(value);
+                self.clipboard_provider.set_contents(
+                    self.inner[&name].join(NATIVE_LINE_ENDING.as_str()),
+                    match name {
+                        '*' => ClipboardType::Clipboard,
+                        '+' => ClipboardType::Selection,
+                        _ => unreachable!(),
+                    },
+                )
+            }
+            _ => {
+                self.inner.entry(name).or_insert_with(Vec::new).push(value);
+                Ok(())
+            }
+        }
+    }
+
+    pub fn first<'a>(&'a self, name: char, editor: &'a Editor) -> Option<Cow<'a, str>> {
+        self.read(name, editor).and_then(|mut values| values.next())
+    }
+
+    pub fn last<'a>(&'a self, name: char, editor: &'a Editor) -> Option<Cow<'a, str>> {
+        self.read(name, editor).and_then(|values| values.last())
+    }
+
+    pub fn iter_preview(&self) -> impl Iterator<Item = (char, &str)> {
+        self.inner
+            .iter()
+            .filter(|(name, _)| !matches!(name, '*' | '+'))
+            .map(|(name, values)| {
+                let preview = values
+                    .last()
+                    .and_then(|s| s.lines().next())
+                    .unwrap_or("<empty>");
+
+                (*name, preview)
+            })
+            .chain(
+                [
+                    ('_', "<empty>"),
+                    ('#', "<selection indices>"),
+                    ('.', "<selection contents>"),
+                    ('%', "<document path>"),
+                    ('*', "<system clipboard>"),
+                    ('+', "<primary clipboard>"),
+                ]
+                .iter()
+                .copied(),
+            )
+    }
+
+    pub fn clear(&mut self) {
+        self.inner.clear()
+    }
+
+    pub fn remove(&mut self, name: char) -> bool {
+        match name {
+            '_' | '#' | '.' | '%' | '*' | '+' => false,
+            _ => self.inner.remove(&name).is_some(),
+        }
+    }
+
+    pub fn clipboard_provider_name(&self) -> Cow<str> {
+        self.clipboard_provider.name()
+    }
+}
+
+fn read_from_clipboard<'a>(
+    provider: &dyn ClipboardProvider,
+    saved_values: Option<&'a Vec<String>>,
+    clipboard_type: ClipboardType,
+) -> RegisterValues<'a> {
+    match provider.get_contents(clipboard_type) {
+        Ok(contents) => {
+            // If we're pasting the same values that we just yanked, re-use
+            // the saved values. This allows pasting multiple selections
+            // even when yanked to a clipboard.
+            let Some(values) = saved_values else { return Box::new(iter::once(contents.into())) };
+
+            if contents_are_saved(values, &contents) {
+                Box::new(values.iter().map(Cow::from))
```

Comment:
```suggestion
                Box::new(values.iter().map(Cow::from).rev())
```

needs to be reversed for consistency (same as the other registers)